### PR TITLE
Dev fused bn functional

### PR DESCRIPTION
**Pull request by MARD1NO**
增加 BatchNorm add Relu 的接口

一些未完成的细节：
1. cpu版本没实现，后续考虑补齐
2. affine = False参数不支持，需要在functional做较大改动，用户一般不会选择这样的选项

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    // (x, addend, gamma, beta) or (x, addend, moving_mean, moving_variance, gamma, beta)
+
+    // ref to track_running_stats false/true
+    // output_size may be 2 or 4, as outputs may be
+    // (x, reserve_space) or (x, reserve_space, mean, inv_variance)
+    // ref to is_training false/true
+    ctx->x_requires_grad = inputs.at(0)->requires_grad();
+    std::shared_ptr<Tensor> add_end, gamma, beta;
+
+    if (inputs.size() == 3 || inputs.size() == 5) {
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp:67]: 这一段用 switch 是不是会好看点

File: oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    ctx->SaveTensorForBackward(beta);          // beta 2
+
+    // input_size may be 3/4/5/6, as inputs may be
+    // (x, gamma, beta) or (x, moving_mean, moving_variance, gamma, beta)
+    // (x, addend, gamma, beta) or (x, addend, moving_mean, moving_variance, gamma, beta)
+
+    // ref to track_running_stats false/true
+    // output_size may be 2 or 4, as outputs may be
+    // (x, reserve_space) or (x, reserve_space, mean, inv_variance)
+    // ref to is_training false/true
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp:113]: 这段注释是不是重复了

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 3]
```diff
@@ -488,6 +488,147 @@ class NormalizationFunctor {
...
+             gamma, beta},
+            attrs);
+      } else {
+        return OpInterpUtil::Dispatch<one::Tensor>(
+            *norm_training_stats_op_,
+            {x, JUST(moving_mean.value()), JUST(moving_variance.value()), gamma, beta}, attrs);
+      }
+    } else {
+      printf("here no no states!");
+
```
Comment (wyg1997) [inline oneflow/core/functional/impl/nn_functor.cpp:610]: 文件里的 print 记得删掉

File: python/oneflow/nn/modules/batchnorm_fused.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            strict,
+            missing_keys,
+            unexpected_keys,
+            error_msgs,
+        )
+
+    def extra_repr(self):
+        return "{num_features}, eps={eps}, momentum={momentum}, affine={affine}, track_running_stats={track_running_stats}".format(
+            **self.__dict__
+        )
```
Comment (wyg1997) [inline python/oneflow/nn/modules/batchnorm_fused.py:92]: 这一块把其它参数也加上吧，对这个 module 的 debug 还挺重要的

File: python/oneflow/nn/modules/batchnorm_fused.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    def _load_from_state_dict(
+        self,
+        state_dict,
+        prefix,
+        local_metadata,
+        strict,
+        missing_keys,
+        unexpected_keys,
+        error_msgs,
+    ):
```
Comment (wyg1997) [inline python/oneflow/nn/modules/batchnorm_fused.py:78]: 重写这个函数有什么特别的用意吗？好像没有对父类的方法做重写

File: python/oneflow/nn/modules/batchnorm_fused.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+            num_features, has_addend, eps, momentum, affine, track_running_stats
+        )
+
+    def forward(self, x, addend):
+        self._check_input_dim(x)
+        # TODO(zwx): Use `tensor.device_type()` method to help checking if x is on cpu.
+        # Using `if x.device == flow.device("cpu"):` will fail as consistent tensor has
+        # no device, however using `x.is_cuda` is not a good choice.
+        # import pdb
+        # pdb.set_trace()
```
Comment (wyg1997) [inline python/oneflow/nn/modules/batchnorm_fused.py:115]: 这个要删掉

File: python/oneflow/nn/modules/batchnorm_fused.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+        if self.has_addend is False:
+            if not x.is_cuda:
+                # can add cpu version.
```
Comment (wyg1997) [inline python/oneflow/nn/modules/batchnorm_fused.py:118]: 这个注释是啥意思

File: oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp (hunk 1 above)
Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/normalization_add_relu.cpp:67]: NOTE：逻辑不方便合并，这里还是用现在的方式来写

File: python/oneflow/test/modules/test_bn_add_relu.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    test_case.assertIsNone(fused_bn.running_mean)
+    test_case.assertIsNone(origin_batch_norm.running_mean)
+    test_case.assertIsNone(fused_bn.running_var)
+    test_case.assertIsNone(origin_batch_norm.running_var)
+
+
+@flow.unittest.skip_unless_1n1d()
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test gpu cases")
+class TestBnAddRelu(flow.unittest.TestCase):
+    def test_gather(test_case):
```
Comment (simonJJJ) [inline python/oneflow/test/modules/test_bn_add_relu.py:288]: 改下函数名字

File: python/oneflow/test/modules/test_bn_add_relu.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        ]
+        arg_dict["channels"] = [4, 6, 8, 10]
+        arg_dict["height"] = [6, 8, 10, 16]
+        arg_dict["width"] = [12, 8, 4, 10]
+        for arg in GenArgList(arg_dict):
+            arg[0](test_case, *arg[1:])
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (simonJJJ) [inline python/oneflow/test/modules/test_bn_add_relu.py:303]: 好像没测1d和3d

**Review by simonJJJ — APPROVED** [review]
